### ISSUE TITLE
Moving to WebExtension (FF 51-57)

### ISSUE THREAD

**Issue by schomery** [schomery/privacy-settings#80]
https://addons.mozilla.org/en-US/firefox/addon/privacy-settings-webextension/

How to install
1. Open about:debugging
2. Download the XPI; https://addons.mozilla.org/firefox/downloads/file/595163/privacy_settings_webextension-0.1.0-fx.xpi?src=dp-btn-primary
3. Point "Load temporary ..." to the downloaded XPI

![screen shot 2017-02-20 at 17 44 06](https://cloud.githubusercontent.com/assets/11704051/23128650/f571ebce-f794-11e6-8a13-899621176737.png)

Notes:
1. There will be no more advanced settings page. The new design allow resizing, so we can have more options here
2. Still tooltip text is not displayed correctly in this WebExtension version. We will have links to the settings description soon

**Comment by Gitoffthelawn** [comment]
I'm trying to understand how pref changes are controlled in WebExtensions.

According to this thread, it is not currently possible:

- https://bitbucket.org/cat_in_136/configuration-mania/issues/254/will-cm-be-converted-to-a-webextension

But you seem to be able to do it! :)

What's the story?

**Comment by yfdyh000** [comment]
@Gitoffthelawn This is an embedded WebExtension, it will not works in Fx57 and above if Mozilla does not provide a new interface.

@schomery The summary should to "(FF 51-56)"?

**Comment by Gitoffthelawn** [comment]
@yfdyh000 Thanks... I'm starting to learn about Embedded WebExtensions.  Kind of confusing at first.  Can you give me a simple explanation as to what they are and why they are used at all?

**Comment by yfdyh000** [comment]
https://developer.mozilla.org/Add-ons/WebExtensions/Embedded_WebExtensions
`privacy_settings_webextension-0.1.0-fx.xpi` looks use it to call the [sdk/preferences/service](https://developer.mozilla.org/Add-ons/SDK/Low-Level_APIs/preferences_service).

**Comment by Gitoffthelawn** [comment]
Thanks.  That's what I've been reading.  Going to have to read it a few times to fully understand it.

**Comment by schomery** [comment]
Yes, we are using SDK side to save and fetch internal preferences. Generally, the WebExtension side is more stable and memory friendly. That's the reason why I decided to use embedded WebExtension approach to rewriting the extension from scratch. If there will be no way to access internal prefs after 57 release, that would be basically the end of this extension for Firefox browser.

**Comment by Gitoffthelawn** [comment]
Hopefully Mozilla will improve the WebExtensions API to handle this before 57 is released.

**Comment by yfdyh000** [comment]
[Bug 1363856 - [tracker] Implement WebExtensions API to give access to specific preferences](https://bugzilla.mozilla.org/show_bug.cgi?id=1363856)

**Comment by CryptOwl** [comment]
Well, it looks like Mozilla is making progress with necessary APIs. What is your take on the situation @schomery ?

**Comment by schomery** [comment]
@CryptOwl what APIs do we have access to?

**Comment by ZatsuneNoMokou** [comment]
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/browserSettings
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/types/BrowserSetting/set
Is it that?

**Comment by schomery** [comment]
@ZatsuneNoMokou these are prototypes. There is no way to alter internal preferences at this point as far as I know.

**Comment by ZatsuneNoMokou** [comment]
I see :/ for now, so WaterFox project still looks like the best alternative for 57+

**Comment by ZatsuneNoMokou** [comment]
https://github.com/ChrisAntaki/disable-webrtc-firefox/blob/master/js/disable_webrtc.js
That addon is able to change a setting, if it can help (this author have like 4 addons changing settings)

**Comment by n8v8R** [comment]
Is the WebExtention going to show soon as addon, considering that FF 57 is not too far off?

**Comment by lightdot** [comment]
Firefox in the Beta channel just disabled all legacy extensions.

**Comment by schomery** [comment]
> Is the WebExtention going to show soon as addon

We still don't have enough settings to play with!

**Comment by kevlar1818** [comment]
@schomery Any updates on the WebExtension version for FF57+? It's still showing up as "legacy".

**Comment by DouglasMoody** [comment]
@schomery Please we need you to retain this corner of the extension market.  Nothing else touches it and I see you just updated your noscript alternative "privacy control" but umatrix already has that covered.  https://www.mozilla.org/en-US/firefox/57.0a1/whatsnew

Could go back to an ESR version but not a good idea. Can you list all the about:config areas your extension edits and we can go a manual method of replicating?

**Comment by y0y0z** [comment]
@schomery: **i really love your work, i would really like to thank you for making many of us happy with your fine add-ons.** although i truly love firefox, i hate the fact that mozilla caused many add-ons to become deprecated.

i tried to install the webextension add-on but it is incompatible with firefox 57.
i have few questions:
1. did you release the add-on for firefox 57 (even in alpha state)?
2. do you intend of making a webextension version of the add-on?
3. are there any alternative recommended by you?
4. besides time, is there any technical difficulties making a webextension version of the add-on?

**Comment by schomery** [comment]
There is still no useful API to change these internal configs. Until Mozilla decides to allow this, there is no way to move to the actual WebExtension.

**Comment by Hasbreak** [comment]
As @DouglasMoody brought in, we need to list all the areas this extension edits and its recommended settings and defaults to set them down on some sort of a cheat sheet for the later use since I and many were heavily dependent on it, and it's a huge loss for us.

**Comment by MWin123** [comment]
@Hasbreak I checked the source code and found the preferences. Feel free to make a cheat sheet.

Suggestions https://github.com/schomery/privacy-settings/blob/eb7b80d50b8f95e4729f966dfa02d3a7bdc0b1e8/src/lib/main.js#L51

All preferences https://github.com/schomery/privacy-settings/blob/eb7b80d50b8f95e4729f966dfa02d3a7bdc0b1e8/src/lib/main.js#L117

There are even more preferences here: https://www.privacytools.io/#about_config Although some of them are rather invasive, e.g. one limits the browser window to a certain resolution.

@yfdyh000 The author wrote the following in the bug tracker that you linked:
> If there is something else you need, please open a separate bug for it, and feel free to make it block this bug.

Does that mean that we have to open a bug report for all missing preferences?

**Comment by TerkiKerel** [comment]
Still not works with latest stable version of Firefox :(

**Comment by IzzySoft** [comment]
@TerkiKerel blame Mozilla for cutting of a weorking API before having its replacement done. If you want a working installation, never update it to a version known to break things. It was clear from the beginning that FF57 will leave many addons in the dark.

**Comment by y0y0z** [comment]
@schomery i came across [First Party Isolation](https://addons.mozilla.org/en-US/firefox/addon/first-party-isolation/) which basically changes the property `privacy.firstparty.isolate` by toggling its boolean value. could it shed some light?

**Comment by schomery** [comment]
@y0y0z Privacy Settings support this. You need to have FF > 58.

![screen shot 2017-11-28 at 5 52 57 pm](https://user-images.githubusercontent.com/11704051/33324500-0e2d403e-d465-11e7-9039-66952b72a02b.png)

**Comment by y0y0z** [comment]
@schomery amazing!

**Comment by TerkiKerel** [comment]
@schomery  Sorry but I could not understand the problem exactly.

- Mozilla deprecated the API to change about-config properties from add ons?
or
- you did not add on the new add on the properties yet? Because I saw on your last screnshot the new design.  I could not understand where other properties gone :/

**Comment by schomery** [comment]
> Mozilla deprecated the API to change about-config properties from add ons?

Yes!
